File: 003_longestSubstringWithoutRepeatingCharacters.cpp
#include <string>

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.empty())
            return 0;
        
        int table[256] = {};
        for (int i=0;i<256;i++) table[i] = -1;
        int p1=0;
        int p2=0;
        int maxLen = 0;
        while (p2 < s.size()) {
            if (table[s[p2]] != -1) {
                maxLen = std::max(maxLen, p2-p1);
                while (p1 < table[s[p2]]) {
                    table[s[p1]] = -1;
                    p1++;
                }
                p1++;
            }
            table[s[p2]] = p2;
            p2++;
        }
        maxLen = std::max(maxLen, p2-p1);
        return maxLen;
    }
};

File: 003_longestSubstringWithoutRepeatingCharacters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s == null || s.length() == 0)
            return 0;
        
        int longestSubStrLen = 0;
        int p1 = 0;
        int p2 = 0;
        Set<Character> set = new HashSet();
        while (p2 < s.length()) {
            if (!set.contains(s.charAt(p2))) {
                set.add(s.charAt(p2));
            } else {
                longestSubStrLen = Math.max(longestSubStrLen, p2-p1);
                while (s.charAt(p1) != s.charAt(p2)) {
                    set.remove(s.charAt(p1));
                    p1++;
                }
                p1++;
            }
            p2++;
        }
        longestSubStrLen = Math.max(longestSubStrLen, p2-p1);
        return p1==0? s.length():longestSubStrLen;
    }
}